### PR TITLE
Introduce Project models, clean-up logic in CodeReviewsController

### PR DESCRIPTION
**Pull request by jamescodeorg**
Follow-up to #46184.

When doing the initial implementation of CodeReviewsController, it seemed like some of the logic could be simplified if we could introduce an ActiveRecord model for Project (now possible since we recently moved the projects table to the dashboard database).  This change adds these models and reworks some of the logic to use these new models.

Additionally, the following functionality was changed:
- Product clarified that a code review is associated with a project (or, more precisely, a channel), and not the script+level.  We still record the script and level that the code review was opened on in case we need this in the future, but it is no longer used as an identifier when retrieving or creating a code review.
- Implemented the correct logic for determining who can view a code review.  Interestingly, this logic only depends on the user and the project associated with the code review (and not the code review itself) so it has been implemented as an ability on the Project object.  It's a bit confusing and I may revisit this later. (For more context on why this wasn't implemented as an ability on CodeReview, see this [slack thread](https://codedotorg.slack.com/archives/C0T10HG6N/p1652391201816919?thread_ts=1652114719.836469&cid=C0T10HG6N).)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/ability.rb [hunk 1]
```diff
@@ -126,13 +126,20 @@ def initialize(user)
...
+        project.owner_id == user.id
       end
       can :edit, CodeReview, user_id: user.id
-      # TODO: teachers and peers should also be able to see the code review
-      can :read, CodeReview, user_id: user.id
+      can :index_code_reviews, Project do |project|
+        # The user can see the code review if one of the following is true:
+        # 1) the user is the project owner
+        # 2) the user is the teacher of the project owner
+        # 3) the user and the project owner are in the same code reivew group
```
Comment (maureensturgeon) [inline dashboard/app/models/ability.rb:138]: Another criteria is that the teacher has turned on code review for the section, do we want to include that here?

**Review by maureensturgeon — APPROVED** [review]
LGTM!

File: dashboard/app/controllers/code_reviews_controller.rb [hunk 2]
```diff
@@ -6,44 +6,36 @@ class CodeReviewsController < ApplicationController
 
   # GET /code_reviews
   # Returns the list of code reviews and associated comments for the given
-  # project (identified by channel id), script, and level.
+  # project (identified by channel id).
   def index
-    params.require([:channelId, :scriptId, :levelId])
+    params.require([:channelId])
```
Comment (molly-moen) [inline dashboard/app/controllers/code_reviews_controller.rb:11]: This means that for a project-backed level progression (where each level in a series uses the same project) all the levels will have the same code review. Is that the intent here?

Comment (jamescodeorg) [reply dashboard/app/controllers/code_reviews_controller.rb:11]: Yes, that was the behavior we settled on after meeting with Mike and Jessie on Monday.

**Review by molly-moen — APPROVED** [review]
Nice!

File: dashboard/app/models/ability.rb (hunk 1 above)
Comment (jamescodeorg) [reply dashboard/app/models/ability.rb:138]: Thanks for catching this.  From our meeting with Mike, I think we want to add the check for if code review is turned on to case 3 above.  I'll pick this up in a future PR.

**Pull request merged**